### PR TITLE
Compiler: -compat-32 flag when building cmo/cma.

### PR DESCRIPTION
**Pull request by hhugo**
`-compat-32` is currently only used when linking a bytecode executable.
It would make sense to use `-compat-32` for any bytecode object (cmo/cma).

### CODE REVIEW

**Comment by alainfrisch** [comment]
Good idea, but I'd propose to share this piece of code (and generalize the exception constructor to take a filename).  At least Bytepackager and Bytelibrarian could call a function in Bytelink.  For sharing with Emitcode, this is less clear.

**Comment by mshinwell** [comment]
@hhugo Can you try to finish this one off?

**Comment by hhugo** [comment]
@alainfrisch, do you prefer the current approach ?

**Comment by mshinwell** [comment]
@alainfrisch Ping

**Comment by alainfrisch** [comment]
Yes, I prefer the current approach.  But introducing a new module just might be a bit too much.  Isn't there a natural place to put this code in an existing module?   If a new module must be created, don't forget the standard copyright/license header.

**Comment by hhugo** [comment]
I've moved the code `emitcode.ml` and rebased on trunk

**Comment by hhugo** [comment]
@alainfrisch, would you mind taking another look at this ?

**Comment by alainfrisch** [comment]
LGTM.  Perhaps add a test case?

**Comment by hhugo** [comment]
test case added

**Comment by shindere** [comment]
This breaks CI becaue tests/tool-ocamlc-compat32 fails, on several machines,
e.g. linux-32 but not only.

@hhugo can you see why and provide a fix?

**Comment by mshinwell** [comment]
@hhugo is going to work on a fix right now.

@damiendoligez The CI slave for ppc64 appears to have OCaml configured in 32-bit mode.  Is that expected?

**Comment by mshinwell** [comment]
(Please don't revert this, we will apply a patch on top)

**Comment by shindere** [comment]
Mark Shinwell (2017/05/05 07:01 -0700):
> @hhugo is going to work on a fix right now.

Great!

> @damiendoligez The CI slave for ppc64 appears to have OCaml configured
> in 32-bit mode.  Is that expected?

We do not have any real ppc 32 bits machine here at Inria.
So to test the ppc32 build we use the ppc64 machine and configure OCaml
to use the -m32 option when calling gcc.

**Comment by dra27** [comment]
The new test is also failing on Windows (I appreciate that AppVeyor results being unavailable for other reasons would have made this difficult to spot).

**Comment by hhugo** [comment]
Can one have access to the test result? what the size of integers on windows 64 ?

File: testsuite/tests/tool-ocamlc-compat32/Makefile [hunk 1]
```diff
@@ -0,0 +1,17 @@
+BASEDIR=../..
+
+compile:
+	printf " ... testing -compat-32"
+	@$(OCAMLC) -compat-32 -c a.ml 2>&1 | xargs echo ${} > test.result
```
Comment (dra27) [inline testsuite/tests/tool-ocamlc-compat32/Makefile:5]: On the basis that tests should be as simple as possible, why is the output piped to `xargs`?

**Comment by hhugo** [comment]
https://github.com/ocaml/ocaml/pull/1161

**Pull request merged**